File: src/main/scala/com/github/kazuhito_m/twitterbattler/primitive/domain/model/battler/BattlerRepository.scala
package com.github.kazuhito_m.twitterbattler.primitive.domain.model.battler

trait BattlerRepository {

  def convertTwitterIdToId(twitterId: String): BattlerIdentifier

  /**
    * 指定されたId(TwitterID)から、ゲームキャラクタデータを作成、同時に保存する。
    */
  def create(identifier: BattlerIdentifier): Battler

  /**
    * Battlerオブジェクトを保存する。すでにある場合は上書きする。
    */
  def registar(battler: Battler): Unit

  /**
    * BattlerオブジェクトをIDで取得する。
    */
  def get(identifier: BattlerIdentifier): Battler

  /**
    * BattlerオブジェクトをIDで削除する。
    */
  def delete(identifier: BattlerIdentifier): Unit

  /**
    * Battlerオブジェクトが存在するかを真偽値で返す。
    */
  def isExists(identifier: BattlerIdentifier): Boolean

  /**
    * 指定したバトラーIDの友達バトラーをランダムに人数指定で取得する。
    */
  def randomFriendBattlers(identifier: BattlerIdentifier, size: Int): List[Battler]

  /**
    * ランダムに一人のバトラーを取得する。
    */
  def randomOneBattler(): Battler

}

File: server/src/main/java/com/github/kazuhito_m/twitterbattler/primitive/application/repository/BattlerRepository.java
package com.github.kazuhito_m.twitterbattler.primitive.application.repository;

import com.github.kazuhito_m.twitterbattler.primitive.domain.model.battler.Battler;
import com.github.kazuhito_m.twitterbattler.primitive.domain.model.battler.BattlerIdentifier;
import com.github.kazuhito_m.twitterbattler.primitive.domain.model.battler.RandomBattlerPickup;

public interface BattlerRepository extends RandomBattlerPickup {
    BattlerIdentifier convertTwitterIdToId(String twitterId);

    /**
     * 指定されたId(TwitterID)から、ゲームキャラクタデータを作成、同時に保存する。
     */
    Battler create(BattlerIdentifier identifier);

    /**
     * Battlerオブジェクトを保存する。すでにある場合は上書きする。
     */
    void register(Battler battler);

    /**
     * BattlerオブジェクトをIDで取得する。
     */
    Battler get(BattlerIdentifier identifier);

    /**
     * BattlerオブジェクトをIDで削除する。
     */
    void delete(BattlerIdentifier identifier);

    /**
     * Battlerオブジェクトが存在するかを真偽値で返す。
     */
    boolean isExists(BattlerIdentifier identifier);
}
